### PR TITLE
[Feature] Add mixmim backbone with inference

### PR DESCRIPTION
**Pull request by wangbo-zhao**
MMselfSelf want to support "[MixMIM: Mixed and Masked Image Modeling for Efficient Visual Representation Learning](https://arxiv.org/abs/2205.13137)". We hope mmcls can support the backbone proposed in it. This PR has aligned the inference precision.
## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/models/backbones/mixmim.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+
+        x = shortcut + self.drop_path(x)
+
+        x = self.ffn(self.norm2(x), identity=x)  # ffn contains DropPath
+
+        return x
+
+
+class MixMIMLayer(BaseModule):
+    """MixMIM Layer. Implements one layer in MixMIM.
```
[the remark this answers is not in the document]
Comment (wangbo-zhao) [reply mmcls/models/backbones/mixmim.py]: done

File: mmcls/models/backbones/mixmim.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+from mmcls.models.backbones.vision_transformer import TransformerEncoderLayer
+from mmcls.models.utils.attention import WindowMSA
+from mmcls.models.utils.helpers import to_2tuple
+from mmcls.registry import MODELS
+
+
+class MixMIMWindowAttention(WindowMSA):
+    """MixMIM Window Attention.
+
+    Implements one windown attention in MixMIM.
```
[the remark this answers is not in the document]
Comment (wangbo-zhao) [reply mmcls/models/backbones/mixmim.py:24]: done

File: configs/mixmim/metafile.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      Parameters: 88344000 # 88M
+      Training Data:
+        - ImageNet-1k
+    Results:
+    - Dataset: ImageNet-1k
+      Task: Image Classification
+      Metrics:
+        Top 1 Accuracy: 84.6
+        Top 5 Accuracy: 97.0
+    Weights: https://download.openmmlab.com/mmclassification/v0/beit/beit-base_3rdparty_in1k_20221114-c0a4df23.pth
```
[the remark this answers is not in the document]
Comment (wangbo-zhao) [reply configs/mixmim/metafile.yml]: done https://download.openmmlab.com/mmclassification/v0/mixmim/mixmim-base_3rdparty_in1k_20221206-e40e2c8c.pth

File: configs/mixmim/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+*Models with * are converted from the [official repo](https://github.com/Sense-X/MixMIM). The config files of these models are only for inference.*
+
+For MixMIM self-supervised learning algorithm, welcome to [MMSelfSup page](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/MixMIM) to get more information.
+
+## How to use it?
+
+### Inference
+
+```python
+python ./tools/test.py configs/mixmim/mixmim-base-p16_8xb64-pt300e-in1k.py  mixmim-base-p16_8xb64-pt300e-in1k_checkpoint.pth
```
[the remark this answers is not in the document]
Comment (wangbo-zhao) [reply configs/mixmim/README.md]: done

File: configs/mixmim/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+<div align=center>
+<img src="https://user-images.githubusercontent.com/56866854/202853730-d26fb3d7-e5e8-487a-aad5-e3d4600cef87.png"/>
+</div>
+
+## Models
+
+|     Model     | Params(M) | Pretrain Epochs | Flops(G) | Top-1 (%) | Top-5 (%) |                      Config                      |         Download          |
+| :-----------: | :-------: | :-------------: | :------: | :-------: | :-------: | :----------------------------------------------: | :-----------------------: |
+| MixMIM-Base\* |    88     |       300       |   16.3   |   84.6    |   97.0    | [config](./mixmim-base-p16_8xb64-pt300e-in1k.py) | [model](<>)  \| [log](<>) |
```
[the remark this answers is not in the document]
Comment (wangbo-zhao) [reply configs/mixmim/README.md]: done. Since this config is only of inference, I delet the log here.

File: configs/mixmim/README.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+## Models
+
+|     Model     | Params(M) | Pretrain Epochs | Flops(G) | Top-1 (%) | Top-5 (%) |                      Config                      |                                         Download                                          |
+| :-----------: | :-------: | :-------------: | :------: | :-------: | :-------: | :----------------------------------------------: | :---------------------------------------------------------------------------------------: |
+| MixMIM-Base\* |    88     |       300       |   16.3   |   84.6    |   97.0    | [config](./mixmim-base-p16_8xb64-pt300e-in1k.py) | [model](https://download.openmmlab.com/mmclassification/v0/mixmim/mixmim-base_3rdparty_in1k_20221206-e40e2c8c.pth) |
+
+*Models with * are converted from the [official repo](https://github.com/Sense-X/MixMIM). The config files of these models are only for inference.*
+
+For MixMIM self-supervised learning algorithm, welcome to [MMSelfSup page](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/MixMIM) to get more information.
```
Comment (fangyixiao18) [inline configs/mixmim/README.md:40]:
```suggestion
For MixMIM self-supervised learning algorithm, welcome to [MMSelfSup page](https://github.com/open-mmlab/mmselfsup/tree/dev-1.x/configs/selfsup/mixmim) to get more information.
```

File: configs/mixmim/metafile.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    Results:
+    - Dataset: ImageNet-1k
+      Task: Image Classification
+      Metrics:
+        Top 1 Accuracy: 84.6
+        Top 5 Accuracy: 97.0
+    Weights: https://download.openmmlab.com/mmclassification/v0/mixmim/mixmim-base_3rdparty_in1k_20221206-e40e2c8c.pth
+    Converted From:
+      Code: https://github.com/Sense-X/MixMIM
+    Config: configs/mixmim/mixmim-base-p16_8xb64-pt300e-in1k.py
```
Comment (fangyixiao18) [inline configs/mixmim/metafile.yml:39]: @mzr1996 not sure about the name convention in mmcls, but I think the name here is not correct

File: mmcls/models/backbones/mixmim.py [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+        drop_path_rate (float): stochastic depth rate. Defaults to 0.
+        attn_drop_rate (float): attention drop rate. Defaults to 0.
+        use_checkpoint (bool): Whether use the checkpoint to
+        reduce GPU memory cost.
+        init_cfg (dict, optional): Initialization config dict.
+            Defaults to None.
+    """
+    arch_settings = {
+        **dict.fromkeys(
+            ['B', 'base'], {
```
Comment (fangyixiao18) [inline mmcls/models/backbones/mixmim.py:359]:
```suggestion
            ['b', 'base'], {
```
small letter of arch, the same as 'l', 'h'

File: tests/test_models/test_backbones/test_mixmim.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+import torch
+
+from mmcls.models.backbones import MixMIMTransformer
+
+
+class TestMixMIM(TestCase):
+
+    def setUp(self):
+        self.cfg = dict(arch='B', drop_rate=0.0, drop_path_rate=0.1)
```
Comment (fangyixiao18) [inline tests/test_models/test_backbones/test_mixmim.py:13]:
```suggestion
        self.cfg = dict(arch='b', drop_rate=0.0, drop_path_rate=0.1)
```

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: configs/mixmim/README.md (hunk 6 above)
Comment (wangbo-zhao) [reply configs/mixmim/README.md]: done

File: configs/mixmim/metafile.yml (hunk 7 above)
Comment (wangbo-zhao) [reply configs/mixmim/metafile.yml:39]: wait @mzr1996 to confirm

File: mmcls/models/backbones/mixmim.py [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+from mmcls.models.utils.helpers import to_2tuple
+from mmcls.registry import MODELS
+
+
+class MixMIMWindowAttention(WindowMSA):
+    """MixMIM Window Attention. Compared with WindowMSA, we add some
+    modifications in ``forward`` to meet the requirement of MixMIM during
+    pretraining.
+
+    Implements one windown attention in MixMIM.
```
Comment (mzr1996) [inline mmcls/models/backbones/mixmim.py:24]:
```suggestion
    """MixMIM Window Attention.
    
    Compared with WindowMSA, we add some modifications
    in ``forward`` to meet the requirement of MixMIM during
    pretraining.
```

File: configs/mixmim/metafile.yml (hunk 7 above)
Comment (mzr1996) [reply configs/mixmim/metafile.yml:39]: Use `mixmim-base_3rdparty_in1k` here.

File: mmcls/models/backbones/mixmim.py [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+    input_resolution={self.input_resolution}, depth={self.depth}'
+
+
+@MODELS.register_module()
+class MixMIMTransformer(BaseBackbone):
+    """MixMIM backbone.
+    A PyTorch implement of : ` MixMIM: Mixed and Masked Image
+    Modeling for Efficient Visual Representation Learning
+    <https://arxiv.org/abs/2205.13137>`_
+    Args:
```
Comment (mzr1996) [inline mmcls/models/backbones/mixmim.py:326]:
```suggestion
    """MixMIM backbone.
    
    A PyTorch implement of `MixMIM: Mixed and Masked Image
    Modeling for Efficient Visual Representation Learning
    <https://arxiv.org/abs/2205.13137>`_
    
    Args:
```

File: mmcls/models/backbones/mixmim.py [hunk 12]
```diff
@@ -0,0 +1,501 @@
...
+        arch (str | dict): MixMIM architecture. If use string,
+            choose from 'base','large' and 'huge'.
+            If use dict, it should have below keys:
+            - **embed_dims** (int): The dimensions of embedding.
+            - **depths** (int): The number of transformer encoder layers.
+            - **num_heads** (int): The number of heads in attention modules.
+            Defaults to 'base'.
```
Comment (mzr1996) [inline mmcls/models/backbones/mixmim.py:333]:
```suggestion
            If use dict, it should have below keys:
            
            - **embed_dims** (int): The dimensions of embedding.
            - **depths** (int): The number of transformer encoder layers.
            - **num_heads** (int): The number of heads in attention modules.
            
            Defaults to 'base'.
```

File: mmcls/models/backbones/mixmim.py [hunk 13]
```diff
@@ -0,0 +1,501 @@
...
+
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            # we use xavier_uniform following official JAX ViT:
+            torch.nn.init.xavier_uniform_(m.weight)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
```
Comment (mzr1996) [inline mmcls/models/backbones/mixmim.py:486]: Set `init_cfg` in the config file or the default value, instead use `init_weights` method.

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: configs/mixmim/metafile.yml (hunk 7 above)
Comment (wangbo-zhao) [reply configs/mixmim/metafile.yml:39]: done

File: mmcls/models/backbones/mixmim.py (hunk 13 above)
Comment (wangbo-zhao) [reply mmcls/models/backbones/mixmim.py:486]: I find that I do not use self._init_weights in MixMIMTransformer anywhere, so I remove it.

File: configs/mixmim/metafile.yml [hunk 14]
```diff
@@ -36,4 +36,4 @@ Models:
     Weights: https://download.openmmlab.com/mmclassification/v0/mixmim/mixmim-base_3rdparty_in1k_20221206-e40e2c8c.pth
     Converted From:
       Code: https://github.com/Sense-X/MixMIM
-    Config: configs/mixmim/mixmim-base-p16_8xb64-pt300e-in1k.py
+    Config: configs/mixmim/mixmim-base_3rdparty_in1k.py
```
Comment (mzr1996) [inline configs/mixmim/metafile.yml:39]: Well, the model (checkpoint) can be `3rdparty`, but the config file won't be `3rdparty`. Please rename it to `mixmim-base_8xb64_in1k.py`

File: mmcls/models/backbones/mixmim.py [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+        x = self.drop_after_pos(x)
+
+        for layer in self.layers:
+            x = layer(x, attn_mask=None)
+
+        x = self.norm(x)
+        x = self.avgpool(x.transpose(1, 2))  # B C 1
+        x = torch.flatten(x, 1)
+
+        return [x]
```
Comment (mzr1996) [inline mmcls/models/backbones/mixmim.py:497]: The return type should be a tuple instead of a list.

File: mmcls/models/backbones/mixmim.py [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+        self.avgpool = nn.AdaptiveAvgPool1d(1)
+        self.num_patches = self.patch_resolution[0] * self.patch_resolution[1]
+        self.absolute_pos_embed = nn.Parameter(
+            torch.zeros(1, self.num_patches, self.embed_dims),
+            requires_grad=False)
+
+        _, self.norm = build_norm_layer(norm_cfg, self.num_features)
+
+    def init_weights(self):
+        super(MixMIMTransformer, self).init_weights()
```
Comment (mzr1996) [inline mmcls/models/backbones/mixmim.py:482]: Please remove it as well.

**Review by mzr1996 — CHANGES_REQUESTED** [review]

**Review by fangyixiao18 — APPROVED** [review]
LGTM

**Review by mzr1996 — APPROVED** [review]
